python/shark_turbine/transforms/quantization/mm_group_quant.py

```diff
@@ -0,0 +1,167 @@
+    %weight_cast = linalg.generic {{
+        indexing_maps = [
+            affine_map<(d0, d1, d2) -> (d0, d1, d2)>, 
+            affine_map<(d0, d1, d2) -> (d0, d1)>, 
+            affine_map<(d0, d1, d2) -> (d0, d1)>, 
+            affine_map<(d0, d1, d2) -> (d0, d1, d2)>], 
+        iterator_types = ["parallel", "parallel", "parallel"] }} 
+        ins(%weight_exp, %scale, %zp : tensor<{k}x{group0}x{group1}x{lowp_type}>, tensor<{k}x{group0}x{element_type}>, tensor<{k}x{group0}x{element_type}>) 
+        outs(%empty_0 : tensor<{k}x{group0}x{group1}x{element_type}>) {{
+    ^bb0(%in: {lowp_type}, %in_1: {element_type}, %in_2: {element_type}, %out: {element_type}):
+        %16 = arith.extui %in : {lowp_type} to i32
+        %17 = arith.uitofp %16 : i32 to {element_type}
+        %18 = arith.subf %17, %in_2 : {element_type}
+        %19 = arith.mulf %18, %in_1 : {element_type}
+        linalg.yield %19 : {element_type}
+    }} -> tensor<{k}x{group0}x{group1}x{element_type}>
+    %cst = arith.constant 0.000000e+00 : {element_type}
+    %empty_1_dyn = tensor.empty(%m, %k) : tensor<?x?x{element_type}>
+    %empty_1 = tensor.cast %empty_1_dyn : tensor<?x?x{element_type}> to tensor<{m}x{k}x{element_type}>
+    %zero_init = linalg.fill ins(%cst : {element_type}) outs(%empty_1 : tensor<{m}x{k}x{element_type}>) -> tensor<{m}x{k}x{element_type}>
+    %result = linalg.generic {{
+        indexing_maps = [
+            affine_map<(d0, d1, d2, d3) -> (d0, d2, d3)>,
+            affine_map<(d0, d1, d2, d3) -> (d1, d2, d3)>, 
+            affine_map<(d0, d1, d2, d3) -> (d0, d1)>], 
+        iterator_types = ["parallel", "parallel", "reduction", "reduction"] }} 
+        ins(%a_exp, %weight_cast : tensor<{m}x{group0}x{group1}x{element_type}>, tensor<{k}x{group0}x{group1}x{element_type}>) 
+        outs(%zero_init : tensor<{m}x{k}x{element_type}>) {{
+    ^bb0(%in: {element_type}, %in_1: {element_type}, %out: {element_type}):
+        %16 = arith.mulf %in, %in_1 : {element_type}
+        %17 = arith.addf %16, %out : {element_type}
+        linalg.yield %17 : {element_type}
+    }} -> tensor<{m}x{k}x{element_type}>
+    return %result : tensor<{m}x{k}x{element_type}>
+  }}
+}}
+"""
+
+
+class MMGroupQuantRewriterPass(Pass):
+    def run(self):
+        globals = self.globals
+        mms = match_children(self.funcs, TransposedMMMatcher(globals, self.builder))
+
+        for mr in mms:
+            if mr.k is None or mr.n is None:
+                continue
+            self.rewrite(mr)
+
+        self.inline()
+        self.cleanup()
+
+    def rewrite(self, mr: TransposedMMResult):
+        none_to_q = lambda x: "?" if x is None else x
+        inline_module_asm = GROUP_MATMUL_TEMPLATE.format(
+            param_name=mr.param_name,
+            lowp_type="i4",
+            m=none_to_q(mr.m),
+            n=none_to_q(mr.n),
+            k=none_to_q(mr.k),
+            n_div=mr.n // 2,
+            group0=32,
+            group1=128,
```

Comment:
Is it easy to make this a pass parameter? This will be important apparently for CPU where we might want to try a group size of 32. As a result it might also be worth making the group size a part of the weight/scale/zp parameter names.

Reply:
Yeah, done.